File: src/com/watnapp/etipitaka/plus/model/History.java
package com.watnapp.etipitaka.plus.model;
import android.provider.BaseColumns;
import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.util.SparseBooleanArray;
import com.watnapp.etipitaka.plus.helper.BookDatabaseHelper;

public class History extends ModelBase {
  private Context context;
  private int id;
  private String keywords;
  private int language;
  private boolean section1;
  private boolean section2;
  private boolean section3;
  private int result1;
  private int result2;
  private int result3;
  private String content;
  private int score;

  public History() {
    super();
  }

  @Override
  public int getId() {
    return id;
  }

  @Override
  public void setId(int id) {
    this.id = id;
  }

  public String getKeywords() {
    return keywords;
  }

  public void setKeywords(String keywords) {
    this.keywords = keywords;
  }

  public BookDatabaseHelper.Language getLanguage() {
    return BookDatabaseHelper.Language.values()[language];
  }

  public void setLanguage(BookDatabaseHelper.Language language) {
    this.language = language.ordinal();
  }

  public boolean isSection1() {
    return section1;
  }

  public void setSection1(boolean section1) {
    this.section1 = section1;
  }

  public boolean isSection2() {
    return section2;
  }

  public void setSection2(boolean section2) {
    this.section2 = section2;
  }

  public boolean isSection3() {
    return section3;
  }

  public void setSection3(boolean section3) {
    this.section3 = section3;
  }

  public void setSections(SparseBooleanArray sections) {
    setSection1(sections != null && sections.get(0, false));
    setSection2(sections != null && sections.get(1, false));
    setSection3(sections != null && sections.get(2, false));
  }

  public void setResults(int[] results) {
    setResult1(results.length > 0 ? results[0] : 0);
    setResult2(results.length > 1 ? results[1] : 0);
    setResult3(results.length > 2 ? results[2] : 0);
  }

  public int getResult1() {
    return result1;
  }

  public void setResult1(int result1) {
    this.result1 = result1;
  }

  public int getResult2() {
    return result2;
  }

  public void setResult2(int result2) {
    this.result2 = result2;
  }

  public int getResult3() {
    return result3;
  }

  public void setResult3(int result3) {
    this.result3 = result3;
  }

  public String getContent() {
    return content;
  }

  public void setContent(String content) {
    this.content = content;
  }

  public int getScore() {
    return score;
  }

  public void setScore(int score) {
    this.score = score;
  }

  @Override
  public void fromCursor(Cursor cursor, Context context) {
    this.id = cursor.getInt(cursor.getColumnIndex(BaseColumns._ID));
    this.keywords = cursor.getString(cursor.getColumnIndex(HistoryTable.HistoryColumns.KEYWORDS));
    this.language = cursor.getInt(cursor.getColumnIndex(HistoryTable.HistoryColumns.LANGUAGE));
    this.section1 = cursor.getInt(cursor.getColumnIndex(HistoryTable.HistoryColumns.SECTION1)) == 1;
    this.section2 = cursor.getInt(cursor.getColumnIndex(HistoryTable.HistoryColumns.SECTION2)) == 1;
    this.section3 = cursor.getInt(cursor.getColumnIndex(HistoryTable.HistoryColumns.SECTION3)) == 1;
    this.result1 = cursor.getInt(cursor.getColumnIndex(HistoryTable.HistoryColumns.RESULT1));
    this.result2 = cursor.getInt(cursor.getColumnIndex(HistoryTable.HistoryColumns.RESULT2));
    this.result3 = cursor.getInt(cursor.getColumnIndex(HistoryTable.HistoryColumns.RESULT3));
    this.score = cursor.getInt(cursor.getColumnIndex(HistoryTable.HistoryColumns.SCORE));
    this.content = cursor.getString(cursor.getColumnIndex(HistoryTable.HistoryColumns.CONTENT));
    this.context = context;
  }

  @Override
  public ContentValues toContentValues() {
    ContentValues values = new ContentValues();
    values.put(HistoryTable.HistoryColumns.KEYWORDS, this.keywords);
    values.put(HistoryTable.HistoryColumns.LANGUAGE, this.language);
    values.put(HistoryTable.HistoryColumns.SECTION1, this.section1);
    values.put(HistoryTable.HistoryColumns.SECTION2, this.section2);
    values.put(HistoryTable.HistoryColumns.SECTION3, this.section3);
    values.put(HistoryTable.HistoryColumns.RESULT1, this.result1);
    values.put(HistoryTable.HistoryColumns.RESULT2, this.result2);
    values.put(HistoryTable.HistoryColumns.RESULT3, this.result3);
    values.put(HistoryTable.HistoryColumns.SCORE, this.score);
    values.put(HistoryTable.HistoryColumns.CONTENT, this.content);
    return values;
  }

  public static History newInstance(Cursor cursor, Context context) {
    History history = new History();
    history.fromCursor(cursor, context);
    return history;
  }


}

File: app/src/main/java/com/watnapp/etipitaka/plus/model/History.kt
package com.watnapp.etipitaka.plus.model

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.provider.BaseColumns
import android.util.SparseBooleanArray
import com.watnapp.etipitaka.plus.helper.BookDatabaseHelper

class History : ModelBase() {
    private var id = 0
    private var keywords: String? = null
    private var language = 0
    private var section1 = false
    private var section2 = false
    private var section3 = false
    private var result1 = 0
    private var result2 = 0
    private var result3 = 0
    private var content: String? = null
    private var score = 0
    private var buddhawaj = false

    override fun getId(): Int = id

    override fun setId(id: Int) {
        this.id = id
    }

    fun getKeywords(): String? = keywords

    fun setKeywords(keywords: String?) {
        this.keywords = keywords
    }

    fun getLanguage(): BookDatabaseHelper.Language = BookDatabaseHelper.Language.values()[language]

    fun setLanguage(language: BookDatabaseHelper.Language) {
        this.language = language.ordinal
    }

    fun isSection1(): Boolean = section1

    fun setSection1(section1: Boolean) {
        this.section1 = section1
    }

    fun isSection2(): Boolean = section2

    fun setSection2(section2: Boolean) {
        this.section2 = section2
    }

    fun isSection3(): Boolean = section3

    fun setSection3(section3: Boolean) {
        this.section3 = section3
    }

    fun setSections(sections: SparseBooleanArray?) {
        setSection1(sections?.get(0, false) ?: false)
        setSection2(sections?.get(1, false) ?: false)
        setSection3(sections?.get(2, false) ?: false)
    }

    fun setResults(results: IntArray) {
        setResult1(results.getOrElse(0) { 0 })
        setResult2(results.getOrElse(1) { 0 })
        setResult3(results.getOrElse(2) { 0 })
    }

    fun getResult1(): Int = result1

    fun setResult1(result1: Int) {
        this.result1 = result1
    }

    fun getResult2(): Int = result2

    fun setResult2(result2: Int) {
        this.result2 = result2
    }

    fun getResult3(): Int = result3

    fun setResult3(result3: Int) {
        this.result3 = result3
    }

    fun getContent(): String? = content

    fun setContent(content: String?) {
        this.content = content
    }

    fun getScore(): Int = score

    fun setScore(score: Int) {
        this.score = score
    }

    fun isBuddhawaj(): Boolean = buddhawaj

    fun setBuddhawaj(buddhawaj: Boolean) {
        this.buddhawaj = buddhawaj
    }

    override fun fromCursor(cursor: Cursor, context: Context) {
        id = cursor.getInt(cursor.getColumnIndexOrThrow(BaseColumns._ID))
        keywords = cursor.getString(cursor.getColumnIndexOrThrow(HistoryTable.HistoryColumns.KEYWORDS))
        language = cursor.getInt(cursor.getColumnIndexOrThrow(HistoryTable.HistoryColumns.LANGUAGE))
        section1 = cursor.getInt(cursor.getColumnIndexOrThrow(HistoryTable.HistoryColumns.SECTION1)) == 1
        section2 = cursor.getInt(cursor.getColumnIndexOrThrow(HistoryTable.HistoryColumns.SECTION2)) == 1
        section3 = cursor.getInt(cursor.getColumnIndexOrThrow(HistoryTable.HistoryColumns.SECTION3)) == 1
        result1 = cursor.getInt(cursor.getColumnIndexOrThrow(HistoryTable.HistoryColumns.RESULT1))
        result2 = cursor.getInt(cursor.getColumnIndexOrThrow(HistoryTable.HistoryColumns.RESULT2))
        result3 = cursor.getInt(cursor.getColumnIndexOrThrow(HistoryTable.HistoryColumns.RESULT3))
        score = cursor.getInt(cursor.getColumnIndexOrThrow(HistoryTable.HistoryColumns.SCORE))
        content = cursor.getString(cursor.getColumnIndexOrThrow(HistoryTable.HistoryColumns.CONTENT))
        buddhawaj = cursor.getInt(cursor.getColumnIndexOrThrow(HistoryTable.HistoryColumns.BUDDHAWAJ)) == 1
    }

    override fun toContentValues(): ContentValues =
        ContentValues().apply {
            put(HistoryTable.HistoryColumns.KEYWORDS, keywords)
            put(HistoryTable.HistoryColumns.LANGUAGE, language)
            put(HistoryTable.HistoryColumns.SECTION1, section1)
            put(HistoryTable.HistoryColumns.SECTION2, section2)
            put(HistoryTable.HistoryColumns.SECTION3, section3)
            put(HistoryTable.HistoryColumns.RESULT1, result1)
            put(HistoryTable.HistoryColumns.RESULT2, result2)
            put(HistoryTable.HistoryColumns.RESULT3, result3)
            put(HistoryTable.HistoryColumns.SCORE, score)
            put(HistoryTable.HistoryColumns.CONTENT, content)
            put(HistoryTable.HistoryColumns.BUDDHAWAJ, buddhawaj)
        }

    companion object {
        @JvmStatic
        fun newInstance(cursor: Cursor, context: Context): History =
            History().apply { fromCursor(cursor, context) }
    }
}
